subscriptions/tests.py

```diff
@@ -1875,6 +1877,25 @@ def test_resend_message_start(self):
         self.assertEqual(str(resend_request.outbound),
                          "c7f3c839-2bf5-42d1-86b9-ccb886645fb4")
 
+    @override_settings(CELERY_TASK_EAGER_PROPAGATES=False)
+    def test_on_failure_sending_tasks(self):
+        """
+        When something fails in a function task using BaseSendMessage as base
+        make sure the SubscriptionSendFailure item is created.
+        """
+
+        # Setup
+        existing = self.make_subscription_audio()
+        self.make_messages_audio(existing.messageset, 3)
+
+        with pytest.raises(ConnectionError) as excinfo:
+            send_next_message(str(existing.id))
+
+        self.assertTrue(
+            str(excinfo.value).find("Failed to establish a new conn") != -1)
+
+        self.assertEqual(SubscriptionSendFailure.objects.all().count(), 1)
```

Comment:
I'm wondering if there's a way that we can test whether or not the `on_failure` handler on the base class interferes with the retry strategy of `autoretry_for`? Or if there's any documentation around that?

Maybe just try running a celery worker locally and confirming that it does still retry a few times before creating the object, if it's difficult to unit test.

Reply:
This is  exactly what this test was for 🙂, the connection error is for the identity lookup that is not mocked. 

It tries 15 times and then creates the `SubscriptionSendFailure`, the on_failure is called each time. 

The value of `self.request.retries` reaches 0 on the last attempt, then we save the failure.

Reply:
Ah I see, I didn't see the retries count check in the base class.